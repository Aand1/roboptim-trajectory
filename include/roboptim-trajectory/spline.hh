// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class Spline declaration.
 */

#ifndef ROBOPTIM_TRAJECTORY_SPLINE_HH
# define ROBOPTIM_TRAJECTORY_SPLINE_HH
# include <roboptim-trajectory/trajectory.hh>

# include <roboptim-trajectory/fwd.hh>

class bspline;

namespace roboptim
{
  class Spline : public Trajectory<4>
  {
  public:
    Spline (size_type, const vector_t&, int, int, int) throw ();
    virtual ~Spline () throw ();

    virtual vector_t operator () (double) const throw ();

    virtual vector_t derivative (double x, size_type order) const throw ();

    virtual jacobian_t variationConfigWrtParam (double t) const throw ();
    virtual jacobian_t variationDerivWrtParam (double t, size_type order)
      const throw ();
    virtual value_type singularPointAtRank (size_type rank) const;
    virtual vector_t derivBeforeSingularPoint (size_type rank, size_type order)
      const;
    virtual vector_t derivAfterSingularPoint (size_type rank, size_type order)
      const;

  private:
    bspline* spline_;
  };
} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_TRAJECTORY_HH
