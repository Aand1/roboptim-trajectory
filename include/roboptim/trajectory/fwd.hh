// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FWD_HH
# define ROBOPTIM_TRAJECTORY_FWD_HH
# include <roboptim/trajectory/sys.hh>

namespace roboptim
{
  template <typename T>
  class FreeTimeTrajectory;

  template <unsigned dorder>
  class Trajectory;

  template <typename T>
  class StateFunction;

  template <typename T>
  class SumCost;

  template <typename T>
  class TrajectoryCost;

  class Spline;
  class SplineLength;
} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_FWD_HH
