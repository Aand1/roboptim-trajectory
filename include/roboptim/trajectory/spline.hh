// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_SPLINE_HH
# define ROBOPTIM_TRAJECTORY_SPLINE_HH
# include <roboptim/trajectory/sys.hh>

# include <roboptim/trajectory/trajectory.hh>
# include <roboptim/trajectory/fwd.hh>

/// \internal
class bspline;

namespace roboptim
{
  /// \addtogroup roboptim_function
  /// @{

  /// \brief Spline trajectory.
  ///
  /// Implement B-Spline.
  class Spline : public Trajectory<4>
  {
  public:
    /// \brief Instantiate a Spline from its definition.
    ///
    /// Instantiate a Spline from its triplet definition:
    /// - time range (\f$t \in [f_min, t_max]\f$): interval on
    ///   which the spline is defined,
    /// - dimension: output space dimension,
    /// - parameters: vector containing concatenated control points.
    ///
    /// The vector parameters has to satisfy the following conditions:
    /// - size should be at least 4,
    /// - there should be at least two control points,
    /// - the vector has to be valid (it has to contain complete control
    ///   points so its size has to be a multiple of the dimension size).
    ///
    /// \param timeRange spline time range
    /// \param dimension spline dimension
    /// \param parameters vector of parameters defining control points
    /// \param name function title
    Spline (interval_t timeRange, size_type dimension,
	    const vector_t& parameters, std::string name = "spline") throw ();

    /// \brief Copy constructor.
    /// \param spline spline that will be copied
    Spline (const Spline& spline) throw ();

    virtual ~Spline () throw ();

    /// \brief Modify spline parameters.
    virtual void setParameters (const vector_t&) throw ();

    virtual jacobian_t variationConfigWrtParam (double t) const throw ();
    virtual jacobian_t variationDerivWrtParam (double t, size_type order)
      const throw ();
    virtual value_type singularPointAtRank (size_type rank) const;
    virtual vector_t derivBeforeSingularPoint (size_type rank, size_type order)
      const;
    virtual vector_t derivAfterSingularPoint (size_type rank, size_type order)
      const;

    ROBOPTIM_IMPLEMENT_CLONE(Spline)

    virtual Trajectory<derivabilityOrder>* resize (interval_t timeRange)
      const throw ()
    {
      return new Spline (timeRange, this->outputSize (), this->parameters ());
    }

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream& o) const throw ();

  protected:
    void impl_compute (result_t&, double) const throw ();
    void impl_derivative (gradient_t& g, double x, size_type order)
      const throw ();

  private:
    /// \brief Convert parameters to internal representation.
    /// \return internal parameter representation
    vector_t makeBSplineVector ();

    /// \brief Number of control points.
    int nbp_;
    /// \brief Pointer to internal spline implementation.
    bspline* spline_;
  };

  /// Example shows Spline use.
  /// \example spline-gradient.cc

  /// @}

} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_TRAJECTORY_HH
