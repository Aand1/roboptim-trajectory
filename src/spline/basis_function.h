#ifndef BASIS_FUNCTION_H
#define BASIS_FUNCTION_H

#include <boost/numeric/ublas/vector.hpp>
#include <boost/numeric/ublas/io.hpp>
#include <boost/numeric/ublas/matrix.hpp>
#include <boost/numeric/ublas/matrix_proxy.hpp>


class basis_function
{
private:
	basis_function(); 

	void define_sizes();

public:
	basis_function(int nb_P_per_fun_in); 

	basis_function(int nb_fun_in, int nb_P_in, int nb_t_m_in, bool grad_of_time_in, int nb_P_per_fun_in); 

	basis_function(int nb_fun_in, int nb_P_in, int nb_t_m_in, bool grad_of_time_in, bool defined_lc_in, bool free_fun_lc_in, int nb_P_per_fun_in); 

	virtual ~basis_function();

	void define_sizes(int nb_fun_in, int nb_P_in, int nb_t_m_in, bool grad_of_time_in);

	void define_sizes(int nb_fun_in, int nb_P_in, int nb_t_m_in, bool grad_of_time_in, bool defined_lc_in, bool free_fun_lc_in);

	void copy(const basis_function &bf);

	int get_nb_call_def_sizes(); //give the number of call to define_sizes, so that other class getting sizes from this class can track if sizes have changed

	void def_parameters(boost::numeric::ublas::matrix<double> *P_in, double T_in);

	void def_parameters(boost::numeric::ublas::matrix<double> *P_in,
		const boost::numeric::ublas::vector<double> &fun0, 
		const boost::numeric::ublas::vector<double> &funf, double T_in);

	inline void check_T(double T_in)
	{
		if (T_in<0) {
			std::cout << "error in class basis_dunction, duration T must be positive, T = " << T_in << std::endl;
		}
	}

	void define_fixed_lc(
		const boost::numeric::ublas::vector<double> &fun0, 
		const boost::numeric::ublas::vector<double> &funf);

	void get_fixed_lc(boost::numeric::ublas::vector<double> &fun0, 
		boost::numeric::ublas::vector<double> &funf) const;

	void define_fixed_time(double T_in);

	void get_fixed_time(double &T_out);

	double get_free_time();

	bool get_defined_lc();

	bool get_free_fun_lc();

	int get_nb_fun();

	int get_nb_P_free(); //number of free parameters for each joint (if limit conditions are defined it is less than nb_P) 

	int get_Nint();

	void save_parameters(const char* file_name, boost::numeric::ublas::matrix<double> *P_in, double T_in);

	void save_parameters(
		const char* file_name,
		boost::numeric::ublas::matrix<double> *P_in,
		const boost::numeric::ublas::vector<double> &fun0, 
		const boost::numeric::ublas::vector<double> &funf, double T_in);

	void load_parameters(const char* file_name, boost::numeric::ublas::matrix<double> *P_out, double &T_out);

	void load_parameters(const char* file_name,
		boost::numeric::ublas::matrix<double> *P_out,
		boost::numeric::ublas::vector<double> &fun0_out, 
		boost::numeric::ublas::vector<double> &funf_out, double &T_out);

	// read all data of a motion, and save all necessary data inside basis_function class for direct trajectory computation
	void load_parameters_and_sizes(const char* file_name,
		boost::numeric::ublas::matrix<double> *P_out,
		boost::numeric::ublas::vector<double> &fun0_out, 
		boost::numeric::ublas::vector<double> &funf_out, double &T_out);

	// read all data of a motion, and save all necessary data inside basis_function class for direct trajectory computation
	void load_parameters_and_sizes(const char* file_name);

	void define_parameters_scale(double P_scale, double T_scale, double funif_scale,
		const boost::numeric::ublas::vector<double> & funs_scale);

	void get_parameters_scale(double &P_scale, double &T_scale, double &funif_scale,
		boost::numeric::ublas::vector<double> &funs_scale);

	void get_parameters_scale(boost::numeric::ublas::vector<double> &x_scale_out);

	void get_scale_for_auto_scale(int &nb_x_scale_out, 
		boost::numeric::ublas::vector<int> &nb_p_x_scale_p_out, 
		boost::numeric::ublas::matrix<int> &p_x_scale_p_out);

	void set_scale_from_auto_scale(boost::numeric::ublas::vector<double> &p_x_scale_in);

	void convert_parameters_P2x(const boost::numeric::ublas::matrix<double> *P_in, const double T_in, double* x_out);

	void convert_parameters_P2x(const boost::numeric::ublas::matrix<double> *P_in,
		const boost::numeric::ublas::vector<double> &fun0, 
		const boost::numeric::ublas::vector<double> &funf, const double T_in, double* x_out);

	void convert_parameters_P2x(const boost::numeric::ublas::matrix<bool> *P_in, const bool T_in, bool* x_out);

	void convert_parameters_P2x(const boost::numeric::ublas::matrix<bool> *P_in,
		const boost::numeric::ublas::vector<bool> &fun0, 
		const boost::numeric::ublas::vector<bool> &funf, const bool T_in, bool* x_out);

	void convert_parameters_x2P(const double* x_in, boost::numeric::ublas::matrix<double> *P_out, double &T_out);

	void convert_parameters_x2P(const double* x_in,
		boost::numeric::ublas::matrix<double> *P_out,
		boost::numeric::ublas::vector<double> &fun0_out, 
		boost::numeric::ublas::vector<double> &funf_out, double &T_out);


	void change_P_parameter_in_x(boost::numeric::ublas::vector<double> &x_in_out, int i, int j, double P_ij_in);

	void change_fun0_parameter_in_x(boost::numeric::ublas::vector<double> &x_in_out, int i, double qi_i_in);

	void change_funf_parameter_in_x(boost::numeric::ublas::vector<double> &x_in_out, int i, double qf_i_in);

	void change_T_parameter_in_x(boost::numeric::ublas::vector<double> &x_in_out, double T_in);


	void change_P_parameter_in_x(boost::numeric::ublas::vector<bool> &x_in_out, int i, int j, bool P_ij_in);

	void change_fun0_parameter_in_x(boost::numeric::ublas::vector<bool> &x_in_out, int i, bool qi_i_in);

	void change_funf_parameter_in_x(boost::numeric::ublas::vector<bool> &x_in_out, int i, bool qf_i_in);

	void change_T_parameter_in_x(boost::numeric::ublas::vector<bool> &x_in_out, bool T_in);


	double get_P_parameter_in_x(const boost::numeric::ublas::vector<double> &x_in, int i, int j) const;

	double get_fun0_parameter_in_x(const boost::numeric::ublas::vector<double> &x_in, int i) const;

	double get_funf_parameter_in_x(const boost::numeric::ublas::vector<double> &x_in, int i) const;

	double get_T_parameter_in_x(const boost::numeric::ublas::vector<double> &x_in) const;


	bool get_P_parameter_in_x(const boost::numeric::ublas::vector<bool> &x_in, int i, int j) const;

	bool get_fun0_parameter_in_x(const boost::numeric::ublas::vector<bool> &x_in, int i)const;

	bool get_funf_parameter_in_x(const boost::numeric::ublas::vector<bool> &x_in, int i) const;

	bool get_T_parameter_in_x(const boost::numeric::ublas::vector<bool> &x_in)const ;


	void unscale_parameters(boost::numeric::ublas::matrix<double> *P, double &T);

	void unscale_parameters(boost::numeric::ublas::matrix<double> *P,
		boost::numeric::ublas::vector<double> &fun0, 
		boost::numeric::ublas::vector<double> &funf, double &T);

	void scale_parameters(boost::numeric::ublas::matrix<double> *P, double &T);

	void scale_parameters(boost::numeric::ublas::matrix<double> *P,
		boost::numeric::ublas::vector<double> &fun0, 
		boost::numeric::ublas::vector<double> &funf, double &T);

	double unscale_hess(double value_in, int x_dep1, int x_dep2);

	double scale_hess(double value_in, int x_dep1, int x_dep2);

	double unscale_grad(double value_in, int x_dep1);

	double scale_grad(double value_in, int x_dep1);

	int get_nb_grad_fun(); //number of gradient components for a function in compressed form
	int get_nb_grad_comp(); //number of compressed gradient components for a variable depending of all functions
	int get_nb_grad_uncomp(); //number of gradient components of any uncompressed data computed from all functions

	void get_rank_grad_fun2comp(boost::numeric::ublas::matrix<int> &rank_grad_fun2comp_in);

	virtual void get_rank_nz_uncomp(boost::numeric::ublas::vector<int> &time_points_in,
		boost::numeric::ublas::vector<int> &funorcomp_type_in,
		boost::numeric::ublas::vector<int> &nb_nz_per_row,
		boost::numeric::ublas::matrix<int> &rank_grad_u2uncomp_out, int nb_t) = 0;

	void joint2grad_dep(boost::numeric::ublas::vector<int> &nb_fun_dep_in,
		boost::numeric::ublas::matrix<int> &fun_dep_in,
		boost::numeric::ublas::vector<int> &nb_grad_dep_out,
		boost::numeric::ublas::matrix<int> &grad_dep_out);

	bool get_if_grad_of_time();

	virtual void calc_fun(boost::numeric::ublas::vector<double> *all_t, boost::numeric::ublas::matrix<double> *fun, int nb_t) = 0;
	virtual void calc_dfun(boost::numeric::ublas::vector<double> *all_t, boost::numeric::ublas::matrix<double> *dfun, int nb_t) = 0;
	virtual void calc_ddfun(boost::numeric::ublas::vector<double> *all_t, boost::numeric::ublas::matrix<double> *ddfun, int nb_t) = 0;

	virtual void calc_fun_grad(boost::numeric::ublas::vector<double> *all_t, 
		boost::numeric::ublas::matrix<double> *fun, 
		boost::numeric::ublas::matrix< boost::numeric::ublas::vector<double> > *fun_grad, int nb_t) = 0;
	virtual void calc_dfun_grad(boost::numeric::ublas::vector<double> *all_t, 
		boost::numeric::ublas::matrix<double> *dfun, 
		boost::numeric::ublas::matrix< boost::numeric::ublas::vector<double> > *dfun_grad, int nb_t) = 0;
	virtual void calc_ddfun_grad(boost::numeric::ublas::vector<double> *all_t, 
		boost::numeric::ublas::matrix<double> *ddfun, 
		boost::numeric::ublas::matrix< boost::numeric::ublas::vector<double> > *ddfun_grad, int nb_t) = 0;

	virtual void uncompress_grad(boost::numeric::ublas::vector<double> *all_t, 
		boost::numeric::ublas::matrix< boost::numeric::ublas::vector<double> > *data_grad_comp, 
		boost::numeric::ublas::matrix< boost::numeric::ublas::vector<double> > *data_grad_uncomp, int nb_t) = 0;

	void comp_dt(boost::numeric::ublas::vector<double> &t,
		boost::numeric::ublas::matrix<double> &data,
		boost::numeric::ublas::matrix<double> &data_dt, int nb_t);


protected:

	//variables    namespace boost::numeric::ublas;
	int nb_fun;
	int nb_P;
	int nb_P_per_fun; //number of parameters of P from which the value of a function depend whatever the time (must be defined in the constructor of the child class)
	int nb_grad_fun; //number of compressed gradient components of each function returned by class basis_function
	int nb_grad_comp; //number of compressed gradient components for a variable depending of all functions
	int nb_grad_uncomp; //number of gradient components of any uncompressed data computed from all functions
	int nb_call_def_sizes; //number of call to define_sizes, so that other class getting sizes from this class can track if sizes have changed
	double T; //superior limit of time values
	double Tint; //duration of an intervalle
	int Nint; //number of intervalles
	int nb_t_m; //Number max of sample in the range of time
	boost::numeric::ublas::matrix<double> *matP; //Matrix of splines coefficients for each actionnors
	boost::numeric::ublas::vector<double> *all_tn; //Vector of samples of time, normalized
	boost::numeric::ublas::vector<int> N; //N[j] number of values of time in [0,j*Tint]
	boost::numeric::ublas::vector<double> fun0_fixed; //initial value of functions in the fixed initial/final case
	boost::numeric::ublas::vector<double> funf_fixed; //initial value of functions in the fixed initial/final case
	double _T_fixed; //saved value of fixed time

	bool grad_of_time; //if the gradient of time is computed
	bool defined_lc; //1 : the limit conditions of fun are given and limit conditions of dfun and ddfun are zero
	bool free_fun_lc;  //the limit conditions of fun are free (1) or fixed (0)
	// (the difference is for the gradient and where the limit values of fun are given)
	bool given_lc; //in the case of fixed limit conditions, if the limit conditions have been defined
	bool given_T; //in the case of fixed time, if the time have been defined

	//parameters for the scaling
	double funif_scale;
	double P_scale;
	double T_scale;
	boost::numeric::ublas::vector<double> funs_scale;

	virtual void def_Nint() = 0; //must define Nint from nb_P

	virtual void solve_P_0(const boost::numeric::ublas::vector<double> &fun0) = 0;

	virtual void solve_P_f(const boost::numeric::ublas::vector<double> &funf) = 0;

	void calc_N(boost::numeric::ublas::vector<double> *all_t, int nb_t); //compute the vector N

	int intervalle_of_time(double t); //give the intervalle of time t

private:


	//member functions


};

#endif // BASIS_FUNCTION_H
