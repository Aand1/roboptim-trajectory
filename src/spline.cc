// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class Spline implentation.
 */

#include <roboptim-trajectory/spline.hh>

#include <spline/bspline.h>

namespace roboptim
{
  Spline::Spline (size_type m, const vector_t& p) throw ()
    : Trajectory<4> (m, p)
  {
  }

  Spline::~Spline () throw ()
  {
  }

  Spline::vector_t
  Spline::derivative (double x, size_type order) const throw ()
  {
    vector_t res (42);
    return res;
  }

  Spline::jacobian_t
  Spline::variationConfigWrtParam (double t) const throw ()
  {
    jacobian_t jac (42, 42);
    return jac;
  }

  Spline::jacobian_t
  Spline::variationDerivWrtParam (double t, size_type order)
    const throw ()
  {
    jacobian_t jac (42, 42);
    return jac;
  }

  Spline::value_type
  Spline::singularPointAtRank (size_type rank) const
  {
    return 42.;
  }

  Spline::vector_t
  Spline::derivBeforeSingularPoint (size_type rank, size_type order) const
  {
    vector_t res (42);
    return res;
  }

  Spline::vector_t
  Spline::derivAfterSingularPoint (size_type rank, size_type order) const
  {
    vector_t res (42);
    return res;
  }
} // end of namespace roboptim.
