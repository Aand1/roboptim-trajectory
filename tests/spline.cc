// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <boost/numeric/ublas/io.hpp>

#include "common.hh"

#include <roboptim-trajectory/fwd.hh>
#include <roboptim-trajectory/spline.hh>

using namespace roboptim;

int run_test ()
{
  Spline::vector_t params (8);

  // Initial position.
  params[0] = 0.,  params[1] = 0.;
  // Control point 1.
  params[2] = 25.,  params[3] = 50.;
  // Control point 2.
  params[4] = 50.,  params[5] = 25.;
  // Final position.
  params[6] = 100., params[7] = 100.;

  Spline spline (std::make_pair (0., 5.), 2, params);

  std::cout << "# Values:" << std::endl;
  std::cout << "# " << spline (0.) << std::endl;
  std::cout << "# " << spline (2.5) << std::endl;
  std::cout << "# " << spline (5.) << std::endl;

  std::cout << "# 1st derivative:" << std::endl;
  std::cout << "# " << spline.derivative (0., 1) << std::endl;
  std::cout << "# " << spline.derivative (2.5, 1) << std::endl;
  std::cout << "# " << spline.derivative (5., 1) << std::endl;

  std::cout << "# 2nd derivative:" << std::endl;
  std::cout << "# " << spline.derivative (0., 2) << std::endl;
  std::cout << "# " << spline.derivative (2.5, 2) << std::endl;
  std::cout << "# " << spline.derivative (5., 2) << std::endl;

  std::cout << "# Start generating GNU plot information...." << std::endl;
  std::cout << "set term x11 enhanced persist" << std::endl;
  std::cout << "plot '-' with line" << std::endl;

  const double step = 0.01;
  for (double t = 0.; t < 5.; t += step)
    {
      Spline::vector_t res = spline (t);
      std::cout << res[0] << " " << res[1] << std::endl;
    }
  std::cout << "e\n" << std::endl;

  return 0;
}

GENERATE_TEST ()
