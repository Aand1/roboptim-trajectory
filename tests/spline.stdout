Spline
Number of discretization points: 20
